pkg/storage/bloom/v1/bloom_tester.go

```diff
@@ -161,38 +163,39 @@ func (sm stringMatcherTest) Matches(bloom filter.Checker) bool {
 		rawCombined = unsafe.Slice(unsafe.StringData(combined), len(combined))
 	)
 
-	if !bloom.Test(rawKey) {
-		// The structured metadata key wasn't indexed. However, sm.matcher might be
-		// checking against a label which *does* exist, so we can't safely filter
-		// out this chunk.
-		//
-		// TODO(rfratto): The negative test here is a bit confusing, and the key
-		// presence test should likely be done higher up within FuseQuerier.
-		return true
-	}
-
-	return bloom.Test(rawCombined)
+	return sm.match(series, bloom, rawKey, rawCombined)
 }
 
-func (sm stringMatcherTest) MatchesWithPrefixBuf(bloom filter.Checker, buf []byte, prefixLen int) bool {
+func (sm stringMatcherTest) MatchesWithPrefixBuf(series labels.Labels, bloom filter.Checker, buf []byte, prefixLen int) bool {
 	var (
 		combined = fmt.Sprintf("%s=%s", sm.matcher.Key, sm.matcher.Value)
 
 		prefixedKey      = appendToBuf(buf, prefixLen, sm.matcher.Key)
 		prefixedCombined = appendToBuf(buf, prefixLen, combined)
 	)
 
-	if !bloom.Test(prefixedKey) {
-		// The structured metadata key wasn't indexed for a prefix. However,
-		// sm.matcher might be checking against a label which *does* exist, so we
-		// can't safely filter out this chunk.
-		//
-		// TODO(rfratto): The negative test here is a bit confusing, and the key
-		// presence test should likely be done higher up within FuseQuerier.
-		return true
+	return sm.match(series, bloom, prefixedKey, prefixedCombined)
+}
+
+func (sm stringMatcherTest) match(series labels.Labels, bloom filter.Checker, key []byte, combined []byte) bool {
```

Comment:
> structured metadata shadows stream labels

You're right. Great catch.

> ```go
> return series.Get(sm.matcher.Key) != sm.matcherValue && bloom.Test(combined)
> ```

We should always check the bloom tho. This expression is not correct since if the series value for env is equal to the matcher value `env` we'd return false (filter chunks).

It should be:
```go
var inSeries bool
if len(series) > 0 {
  val := series.Get(sm.matcher.Key)
  inSeries = val != "" && val == sm.matcher.Value
}

inBloom := bloom.Test(combined)
return inSeries || inBloom
```

I also tried to support the case that structured metadata overrides the series value:
```
series: {env="prod", app="fake"}
bloom: {trace_id="exists_1", trace_id="exists_2", app="other"} # app overriding series
query: `{app="fake"} | app="fake"` 
```
The query will return no data, but we cannot skip it here, since we'd need to know if app is structured metadata. We cannot know that for sure using `bloom.Test("app")` because blooms are probabilistic. Doing this would break checks for stream labels; e.g. bloom.Test("env") may return true even though env is not structured metadata.

We should be able to support this in the future if we have the structured metadata fields in the block or in TSDB.

Reply:
Yes, thank you. I should've inverted the result of my truth table to "keep" instead of "filter out," since the way I had it before threw me off: 

| Is label? | In bloom? | Keep? |
| --------- | --------- | ----- |
| NO        | NO        | NO    |
| YES       | NO        | YES   |
| NO        | YES       | YES   |
| YES       | YES       | YES   | 


So the logical OR is right here.